File: C++/Medium/TimeLimitExceeded/number-of-ways-where-square-of-number-is-equal-to-product-of-two-numbers_8.cpp
class Solution {
public:
    int numTriplets(vector<int>& nums1, vector<int>& nums2) {
        
        int n1 = nums1.size(), n2 = nums2.size();
        int ans = 0;
        for(int i = 0; i < n1; i++){
            for(int j = i+1; j < n1; j++){
                long long prod = (long long)nums1[i]*(long long)nums1[j];
                for(int k = 0; k < n2; k++){
                    if(prod == (long long)nums2[k]*(long long)nums2[k]){
                        ans++;
                    }
                }
            }
        }
        for(int i = 0; i < n2; i++){
            for(int j = i+1; j < n2; j++){
                long long prod = (long long)nums2[i]*(long long)nums2[j];
                for(int k = 0; k < n1; k++){
                    if(prod == (long long)nums1[k]*(long long)nums1[k]){
                        ans++;
                    }
                }
            }
        }
        return ans;
    }
};

File: Java/Medium/Accepted/number-of-ways-where-square-of-number-is-equal-to-product-of-two-numbers_8.java
class Solution {
    public int numTriplets(int[] nums1, int[] nums2) {
        
        return count(nums1, nums2) + count(nums2, nums1);
    }
    private int count(int[] nums1, int[] nums2) {
        int res = 0;
        Map<Long, Integer> map = new HashMap<>();
        for (int i = 0; i < nums1.length; i++) {
            long n = (long) nums1[i] * nums1[i];
            map.put(n, map.getOrDefault(n, 0) + 1);
        }
        for (int i = 0; i < nums2.length; i++) {
            for (int j = i + 1; j < nums2.length; j++) {
                long n = (long) nums2[i] * nums2[j];
                if (map.containsKey(n)) {
                    res += map.get(n);
                }
            }
        }
        return res;
    }
}